### DISCUSSION TITLE
[已解决]在REALITY里，使用 H2 与使用 gRPC 有什么大区别吗？

### DISCUSSION THREAD

**Discussion by chika0801** [XTLS/Xray-core#1719] [Q&A]
在REALITY里，gRPC 比 H2 除了具有H2的特性，还有什么用途上的不一样吗？

在以前时代，gRPC出了有不少人用，基本没人用h2。gRPC后来也发展到过CDN，也有人用。

**Comment by chika0801** [comment]
搜索了下 https://hengyun.tech/thinking-about-grpc-http2/

xray里的gRPC应该是 gRPC over HTTP/2 吧。所以不清楚 gRPC 在 这比H2 多了什么大特性吗？

**Comment by RPRX** [comment]
gun（gRPC）最初就是 @DuckSoft 看到 CloudFlare 支持 gRPC 回源后写的，不是“gRPC后来也发展到过CDN”。

REALITY 不能过免费 CDN，故 gRPC 与 H2 区别不大，由于 gRPC 是 over H2，**直接用 H2 相对省一点点**。
REALITY 支持 gRPC 是顺手写的，just for fun，~~毕竟相比于 H2 大家更喜欢 gRPC，多 padding 一点可能还是好事？~~

你可以看到 Xray-core 内 REALITY 的第一个 commit 就有 REALITY H2 客户端支持，本来是没打算支持 gRPC 的。
~~但是 REALITY WS 就算了吧，这个组合属实没有必要。~~

**Comment by bash99** [comment]
单纯看之前的协议，vless over h2和vless over grpc，后者似乎自带mux效果，线路正常情况下，后续请求握手速度极快，总体页面加载时间有明显感知的提升，甚至超过tuic这类实现了0-RTT模式的协议。

现在REALITY 后续也打算实现0-RTT，但是mux这部分不知道是什么情况。

**Reply by RPRX** [reply]
H2 与 QUIC 都有多路复用，但 H2 基于 TCP，有队头阻塞。QUIC 基于 UDP，没有该问题，但它暂未普及，可能会被运营商 QoS 策略限流。
REALITY 本身专注于安全与隐蔽，想要 0-RTT，加 MUX 或 H2 或 gRPC 就行了，当然还有其它一些方式可以实现 0-RTT，主要给 XTLS 用。

**Reply by bash99** [reply]
之前同是H2和GRPC，应该底层都是tcp，但是多路复用效果差异很大，h2阻塞几率远远高于grpc。

**Reply by RPRX** [reply]
可能 H2 的确有些问题，我看了下代码，H2 是 H2 自带的多路复用，~~还要一个 RTT~~，gRPC 是自己的多路复用，而且是 0-RTT

**Reply by RPRX** [reply]
https://github.com/XTLS/Xray-core/commit/6526e74d49eec5a4bb9a5e448a271d17f262d64b 把 H2 改成了真 0-RTT，有待测试

**Reply by chika0801** [reply]
测试中，一切正常，无比丝滑

**Reply by RPRX** [reply]
> 测试中，一切正常，无比丝滑

好的，测试是 0-RTT 了吗

**Reply by chika0801** [reply]
> 好的，测试是 0-RTT 了吗

不会测怎么样是0-RTT。我用v2rayN加载的自定义配置。以前知道的方法是在v2rayNG上，点测试速度，看延时是多少ms，我想0-RTT的ms延时应该小，但是现在ng还没更新。不会其它测的方法。

再想请问下v2rayNG里的测试多少ms怎么看是几RTT，例如我用vision协议，我到VPS的单面延时是70ms，我点NG里的测试，显示的延时平均在470-500ms。我一直没想明白这算几个RTT。 140ms是一个RTT，X3就520ms快3个RTT了。

如果NG能用H2，0-RTT生效，这个测试结果值应该在 140ms X2 280ms 来判断是生效了的吗？

肯请科普一下，谷歌搜了下相关，一直没搞明白NG里的延时怎么加的。

**Reply by RPRX** [reply]
NG 现在是用一个 HTTPS 请求测延迟：代理 TCP + 代理 TLS + 载荷 TLS + 载荷 HTTP = 4 RTT。旧版 NG 用的是 HTTP 请求，少一个。

**Reply by RPRX** [reply]
你可以用最新 NG 连电脑 REALITY H2，测出 2-RTT 就说明 REALITY H2 自身已经做到了 0-RTT，即和 REALITY gRPC 的延迟一样。

**Reply by chika0801** [reply]
感谢您的解答，我也去问了chatgpt，ping的结果和rtt关系发现自己一直理解错ping的值表示的意思了。我已经看懂你上面的回答了。

**Reply by RPRX** [reply]
我看到你在群里发的测试结果了，transport H2 自身现在是真 0-RTT 了（代理 HTTP 请求的延迟和 ping 一样）

**Comment by H1JK** [comment]
What I want to add is that in Clash the H2 transport doesn't do multiplexing, but gRPC does. Clash opens a new H2 connection for each new inbound connection. So if you're a Clash user and need multiplexing capabilities, consider gRPC first, no matter what proxy protocol you are using.

This is because Clash does not implement H2 transport correctly, and it does not mean that there are problems with H2 itself.

**Reply by bash99** [reply]
wow, Yes I 'm use clash and find grpc is fast that h2. BTW, I've bad experience with h2 when use old v2ray, so maybe it's cause mux.cool has many problems.

**Reply by H1JK** [reply]
@bash99 H2 comes with multiplexing design, not based on Mux.Cool.

**Reply by peytonyip** [reply]
I wonder if the clash mentioned above here includes clash.meta?

**Reply by H1JK** [reply]
@peytonyip Yes currently it includes Clash.Meta. Meta has not refactored H2 yet, but may in the future.
